File: craft/commands/firewall.py
import click

from craft.client import delete, get, post, put
from craft.output import print_item, print_json, print_success, print_table


@click.command("list")
@click.argument("vm_id", required=False, default=None)
def fw_list(vm_id):
    """List firewall rules for a VM."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    data = get(f"/vms/{vm_id}/firewall")
    inner = data.get("data", data)
    rules = inner.get("rules", inner) if isinstance(inner, dict) else inner
    if isinstance(rules, list) and rules:
        rows = []
        for i, r in enumerate(rules):
            rows.append([
                i,
                r.get("type", ""),
                r.get("action", ""),
                r.get("proto", "*"),
                r.get("dport", "*"),
                r.get("source", "*"),
                r.get("enable", 1),
                r.get("comment", ""),
            ])
        print_table(rows, ["#", "Type", "Action", "Proto", "DPort", "Source", "Enabled", "Comment"])
    else:
        print_json(data)


@click.command("add")
@click.argument("vm_id", required=False, default=None)
@click.option("--type", "rule_type", default=None, type=click.Choice(["in", "out"]))
@click.option("--action", default=None, type=click.Choice(["ACCEPT", "DROP", "REJECT"]))
@click.option("--proto", default=None, type=click.Choice(["tcp", "udp", "icmp"]))
@click.option("--dport", default=None, help="Destination port (e.g. 80 or 8000:9000)")
@click.option("--sport", default=None, help="Source port")
@click.option("--source", default=None, help="Source CIDR (e.g. 10.0.0.0/8)")
@click.option("--dest", default=None, help="Destination CIDR")
@click.option("--comment", default=None, help="Comment (max 255 chars)")
@click.option("--enable/--disable", default=True, help="Enable rule (default: enabled)")
def fw_add(vm_id, rule_type, action, proto, dport, sport, source, dest, comment, enable):
    """Add a firewall rule."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()

    # Interactive mode if required fields missing
    if not rule_type or not action:
        from craft.interactive import select_firewall_action
        fw = select_firewall_action()
        rule_type = rule_type or fw["type"]
        action = action or fw["action"]
        proto = proto or fw.get("proto")
        dport = dport or fw.get("dport")
        source = source or fw.get("source")
        comment = comment or fw.get("comment")

    body = {"type": rule_type, "action": action, "enable": 1 if enable else 0}
    if proto:
        body["proto"] = proto
    if dport:
        body["dport"] = dport
    if sport:
        body["sport"] = sport
    if source:
        body["source"] = source
    if dest:
        body["dest"] = dest
    if comment:
        body["comment"] = comment

    data = post(f"/vms/{vm_id}/firewall", body)
    print_success("Firewall rule added.")
    print_item(data)


@click.command("delete")
@click.argument("vm_id", required=False, default=None)
@click.argument("position", required=False, default=None, type=int)
def fw_delete(vm_id, position):
    """Delete a firewall rule by position (0-based)."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    if position is None:
        position = click.prompt("Rule position (0-based)", type=int)
    delete(f"/vms/{vm_id}/firewall/{position}")
    print_success(f"Firewall rule {position} deleted.")


@click.command("options")
@click.argument("vm_id", required=False, default=None)
@click.option("--enable/--disable", default=None, help="Enable or disable firewall")
@click.option("--policy-in", type=click.Choice(["ACCEPT", "DROP", "REJECT"]), default=None)
@click.option("--policy-out", type=click.Choice(["ACCEPT", "DROP", "REJECT"]), default=None)
def fw_options(vm_id, enable, policy_in, policy_out):
    """Update firewall options."""
    if not vm_id:
        from craft.interactive import select_vm
        vm_id = select_vm()
    body = {}
    if enable is not None:
        body["enable"] = 1 if enable else 0
    if policy_in:
        body["policy_in"] = policy_in
    if policy_out:
        body["policy_out"] = policy_out

    if not body:
        click.echo("No options to update. Use --help.")
        return

    put(f"/vms/{vm_id}/firewall/options", body)
    print_success("Firewall options updated.")

File: cmd/firewall.go
package cmd

import (
	"fmt"
	"os"
	"strconv"

	"github.com/AlecAivazis/survey/v2"
	"github.com/CraftIntertech/craft-cli/internal/client"
	"github.com/CraftIntertech/craft-cli/internal/output"
	"github.com/spf13/cobra"
)

func init() {
	// firewall list
	firewallListCmd := &cobra.Command{
		Use:   "list [VM_ID]",
		Short: "List firewall rules for a VM",
		Args:  cobra.MaximumNArgs(1),
		Run:   firewallListRun,
	}

	// firewall add
	firewallAddCmd := &cobra.Command{
		Use:   "add [VM_ID]",
		Short: "Add a firewall rule to a VM",
		Args:  cobra.MaximumNArgs(1),
		Run:   firewallAddRun,
	}
	firewallAddCmd.Flags().String("type", "", "Rule type: in or out")
	firewallAddCmd.Flags().String("action", "", "Rule action: ACCEPT, DROP, or REJECT")
	firewallAddCmd.Flags().String("proto", "", "Protocol: tcp, udp, or icmp")
	firewallAddCmd.Flags().String("dport", "", "Destination port")
	firewallAddCmd.Flags().String("sport", "", "Source port")
	firewallAddCmd.Flags().String("source", "", "Source CIDR")
	firewallAddCmd.Flags().String("dest", "", "Destination CIDR")
	firewallAddCmd.Flags().String("comment", "", "Rule comment")
	firewallAddCmd.Flags().Bool("enable", false, "Enable the rule")
	firewallAddCmd.Flags().Bool("disable", false, "Disable the rule")

	// firewall delete
	firewallDeleteCmd := &cobra.Command{
		Use:   "delete [VM_ID] [POSITION]",
		Short: "Delete a firewall rule",
		Args:  cobra.MaximumNArgs(2),
		Run:   firewallDeleteRun,
	}

	// firewall options
	firewallOptionsCmd := &cobra.Command{
		Use:   "options [VM_ID]",
		Short: "Set firewall options for a VM",
		Args:  cobra.MaximumNArgs(1),
		Run:   firewallOptionsRun,
	}
	firewallOptionsCmd.Flags().Bool("enable", false, "Enable firewall")
	firewallOptionsCmd.Flags().Bool("disable", false, "Disable firewall")
	firewallOptionsCmd.Flags().String("policy-in", "", "Input policy (ACCEPT, DROP, REJECT)")
	firewallOptionsCmd.Flags().String("policy-out", "", "Output policy (ACCEPT, DROP, REJECT)")

	FirewallCmd.AddCommand(firewallListCmd)
	FirewallCmd.AddCommand(firewallAddCmd)
	FirewallCmd.AddCommand(firewallDeleteCmd)
	FirewallCmd.AddCommand(firewallOptionsCmd)
}

func firewallListRun(cmd *cobra.Command, args []string) {
	vmID := ""
	if len(args) > 0 {
		vmID = args[0]
	} else {
		vmID = SelectVM("Select a VM:")
	}

	path := fmt.Sprintf("/vms/%s/firewall", vmID)
	result := client.Get(path, nil, true)

	data, ok := result["data"].([]interface{})
	if !ok {
		fmt.Fprintln(os.Stderr, "No firewall rules found.")
		return
	}

	var rows [][]string
	for i, item := range data {
		rule, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		rows = append(rows, []string{
			strconv.Itoa(i + 1),
			fmt.Sprintf("%v", rule["type"]),
			fmt.Sprintf("%v", rule["action"]),
			fmt.Sprintf("%v", rule["proto"]),
			fmt.Sprintf("%v", rule["dport"]),
			fmt.Sprintf("%v", rule["source"]),
			fmt.Sprintf("%v", rule["enable"]),
			fmt.Sprintf("%v", rule["comment"]),
		})
	}

	output.PrintTable(rows, []string{"#", "Type", "Action", "Proto", "DPort", "Source", "Enabled", "Comment"})
}

func firewallAddRun(cmd *cobra.Command, args []string) {
	vmID := ""
	if len(args) > 0 {
		vmID = args[0]
	} else {
		vmID = SelectVM("Select a VM:")
	}

	ruleType, _ := cmd.Flags().GetString("type")
	action, _ := cmd.Flags().GetString("action")
	proto, _ := cmd.Flags().GetString("proto")
	dport, _ := cmd.Flags().GetString("dport")
	sport, _ := cmd.Flags().GetString("sport")
	source, _ := cmd.Flags().GetString("source")
	dest, _ := cmd.Flags().GetString("dest")
	comment, _ := cmd.Flags().GetString("comment")
	enableFlag, _ := cmd.Flags().GetBool("enable")
	disableFlag, _ := cmd.Flags().GetBool("disable")

	// Interactive prompts for required fields if missing.
	if ruleType == "" {
		var selected string
		prompt := &survey.Select{
			Message: "Select rule type:",
			Options: []string{"in", "out"},
		}
		if err := survey.AskOne(prompt, &selected); err != nil {
			fmt.Fprintf(os.Stderr, "Prompt cancelled: %v\n", err)
			os.Exit(1)
		}
		ruleType = selected
	}

	if action == "" {
		var selected string
		prompt := &survey.Select{
			Message: "Select action:",
			Options: []string{"ACCEPT", "DROP", "REJECT"},
		}
		if err := survey.AskOne(prompt, &selected); err != nil {
			fmt.Fprintf(os.Stderr, "Prompt cancelled: %v\n", err)
			os.Exit(1)
		}
		action = selected
	}

	body := map[string]interface{}{
		"type":   ruleType,
		"action": action,
	}
	if proto != "" {
		body["proto"] = proto
	}
	if dport != "" {
		body["dport"] = dport
	}
	if sport != "" {
		body["sport"] = sport
	}
	if source != "" {
		body["source"] = source
	}
	if dest != "" {
		body["dest"] = dest
	}
	if comment != "" {
		body["comment"] = comment
	}
	if enableFlag {
		body["enable"] = true
	} else if disableFlag {
		body["enable"] = false
	} else {
		body["enable"] = true
	}

	path := fmt.Sprintf("/vms/%s/firewall", vmID)
	client.Post(path, body, true, 0)

	output.PrintSuccess("Firewall rule added successfully.")
}

func firewallDeleteRun(cmd *cobra.Command, args []string) {
	vmID := ""
	position := ""

	switch len(args) {
	case 2:
		vmID = args[0]
		position = args[1]
	case 1:
		vmID = args[0]
		position = PromptText("Enter rule position to delete:")
	default:
		vmID = SelectVM("Select a VM:")
		position = PromptText("Enter rule position to delete:")
	}

	if !Confirm(fmt.Sprintf("Delete firewall rule at position %s?", position)) {
		output.PrintInfo("Cancelled.")
		return
	}

	path := fmt.Sprintf("/vms/%s/firewall/%s", vmID, position)
	client.Delete(path, true)

	output.PrintSuccess("Firewall rule deleted successfully.")
}

func firewallOptionsRun(cmd *cobra.Command, args []string) {
	vmID := ""
	if len(args) > 0 {
		vmID = args[0]
	} else {
		vmID = SelectVM("Select a VM:")
	}

	enableFlag, _ := cmd.Flags().GetBool("enable")
	disableFlag, _ := cmd.Flags().GetBool("disable")
	policyIn, _ := cmd.Flags().GetString("policy-in")
	policyOut, _ := cmd.Flags().GetString("policy-out")

	body := map[string]interface{}{}

	if enableFlag {
		body["enable"] = true
	} else if disableFlag {
		body["enable"] = false
	}
	if policyIn != "" {
		body["policy_in"] = policyIn
	}
	if policyOut != "" {
		body["policy_out"] = policyOut
	}

	if len(body) == 0 {
		fmt.Fprintln(os.Stderr, "No options specified. Use --enable/--disable, --policy-in, or --policy-out.")
		os.Exit(1)
	}

	path := fmt.Sprintf("/vms/%s/firewall/options", vmID)
	client.Put(path, body, true)

	output.PrintSuccess("Firewall options updated successfully.")
}
